services/httpd/msgp_emitter.go

```diff
@@ -0,0 +1,160 @@
+package httpd
+
+import (
+	"io"
+	"net/http"
+
+	"github.com/influxdata/influxdb/influxql"
+	"github.com/tinylib/msgp/msgp"
+)
+
+type messagePackEncoder struct {
+	Epoch     string
+	ChunkSize int
+}
+
+func (e *messagePackEncoder) ContentType() string {
+	return "application/x-msgpack"
+}
+
+func (e *messagePackEncoder) Encode(w io.Writer, header ResponseHeader, results <-chan *influxql.ResultSet) {
+	var convertToEpoch func(row *influxql.Row)
+	if e.Epoch != "" {
+		convertToEpoch = epochConverter(e.Epoch)
+	}
+	values := make([][]interface{}, 0, e.ChunkSize)
+
+	enc := msgp.NewWriter(w)
+	enc.WriteMapHeader(1)
+	enc.WriteString("results")
+	enc.WriteInt(header.Results)
```

Comment:
There's two advantages to having a count in the protocol. The first is that the client knows what to expect. So if you are trying to create a slice of all of the points, I know before reading any points how many are going to be in the current chunk and I also know if I'm expecting more. I can allocate an array that's appropriate.

The second advantage is there is less guessing in the protocol. If you have some kind of signal at the end that tells the stream if it has finished reading, then you have to determine what that signal is and try to find it. It adds an additional condition to the code and requires that you read at least one byte that you then have to buffer just in case that byte is not the EOF signal. If you use an actual EOF signal, that isn't very kind to reusing a socket connection. This is made more difficult because we would need multiple EOF signals. One for the end of a series, another for the end of a result, and another for an end to the response itself.

Since we know what we are going to send before we send it, the first seems less complicated from the client side and has fewer conditions.